scilla/src/backend_collector.rs

```diff
@@ -0,0 +1,365 @@
+use std::collections::HashMap;
+
+use evm::backend::Backend;
+use evm_ds::protos::evm_proto::{Apply, EvmResult, Storage};
+use primitive_types::{H160, H256, U256};
+use serde_json::Value;
+use sha3::{Digest, Keccak256};
+use tracing::*;
+
+pub type Address = H160;
+type AccountStorage = HashMap<H256, H256>;
+
+/// The backend collector acts as a cache during the scilla execution. It responds to queries about the state
+/// and saves changes to the state. Once the execution is complete, it returns the state changes as an EvmResult
+/// which can be applied in the same manner as the EVM flow.
+/// The interface it exposes provides read and write of (key, value), where the key is a string and the value is
+/// an unlimited length byte array.
+/// Due to the way the EVM works, in which the value is always 32 bytes, we pack and unpack the data
+/// into a number of hashes in a way which is hidden from scilla.
+
+#[derive(Debug, Clone)]
+pub struct Account {
+    pub nonce: u64,
+    pub code: Vec<u8>,
+    pub storage_root: Option<H256>,
+    pub is_scilla: bool,
+}
+
+// Structure that answers queries about the state by using the backend, while also collecting
+// the state changes so it can generate an evm result
+pub struct BackendCollector<'a, B: evm::backend::Backend> {
```

Comment:
They do share the same backend in exec.rs

```
            result = if call_args.is_scilla {
                run_scilla_impl_direct(call_args.clone(), &backend)
            } else {
                run_evm_impl_direct(call_args.clone(), &backend)
            };

            if print_enabled {
                debug!("Evm invocation complete - applying result {:?}", result);
            }

            // Apply the results to the backend so they can be used in the next continuation
            backend.apply_update(result.take_apply());
```

Perhaps it would help to rename the scilla one, its purpose is more for caching and compress/decompress. Even if you used the same structure, the scilla one needs additional functionality on top (plus it would be a circular reference). I have tried to keep any ugliness in the scilla side...